src/bin/contribute.rs

```diff
@@ -382,65 +403,129 @@ impl Contribute {
                 .await;
             }
         });
-        for i in 0..total_tasks {
-            let delay_duration = Duration::seconds(DELAY_AFTER_ERROR_DURATION_SECS).to_std()?;
-            let mut cloned = self.clone_with_new_filenames(i);
-            let progress_bar_for_thread = progress_bar.clone();
-            let jh = tokio::spawn(async move {
-                loop {
-                    let result = cloned.run::<E>().await;
-                    if EXITING.load(SeqCst) {
-                        return;
-                    }
-                    match result {
-                        Ok(_) => {}
-                        Err(e) => {
-                            warn!("Got error from run: {}, retrying...", e);
-                            progress_bar_for_thread
-                                .println(&format!("Got error from run: {}, retrying...", e));
-                            if let Some(chunk_id) = cloned.chosen_chunk_id.as_ref() {
-                                if cloned
-                                    .remove_chunk_id_from_lane_if_exists(
-                                        &PipelineLane::Download,
-                                        &chunk_id,
-                                    )
-                                    .expect("Should have removed chunk ID from lane")
-                                {
-                                    let _ =
-                                        cloned.unlock_chunk(&chunk_id, Some(e.to_string())).await;
-                                }
-                                if cloned
-                                    .remove_chunk_id_from_lane_if_exists(
-                                        &PipelineLane::Process,
-                                        &chunk_id,
-                                    )
-                                    .expect("Should have removed chunk ID from lane")
-                                {
-                                    let _ =
-                                        cloned.unlock_chunk(&chunk_id, Some(e.to_string())).await;
-                                }
-                                if cloned
-                                    .remove_chunk_id_from_lane_if_exists(
-                                        &PipelineLane::Upload,
-                                        &chunk_id,
-                                    )
-                                    .expect("Should have removed chunk ID from lane")
-                                {
-                                    let _ =
-                                        cloned.unlock_chunk(&chunk_id, Some(e.to_string())).await;
+
+        loop {
```

Comment:
I've been testing this on a VM that's limited to 192kbps, and so times out often.
I reached a situation where I only have 1 chunk in the Download pipeline and it's stuck there.
Re-reading the code, I believe I understand what happens - only 2/3 async tasks exited, and so the outer loop is not invoked again.

Reply:
If it panics, it should always go to the outer loop. Other errors are handled in the inner loop. Some times will get stuck if the locks are not released on the server (the check for this is ran every 5 minutes), perhaps that's what happened?

Reply:
You're right! I'll give a try to the updated code (if you can commit a version without the crashes it'd be nice)